cli/src/commands/status.rs

```diff
@@ -212,3 +218,60 @@ pub(crate) fn cmd_status(
 
     Ok(())
 }
+
+fn print_collapsed_untracked_files(
+    formatter: &mut dyn Formatter,
+    untracked_paths: &BTreeMap<RepoPathBuf, UntrackedReason>,
+    tree: &MergedTree,
+    path_converter: &RepoPathUiConverter,
+) -> io::Result<()> {
+    formatter.with_label("diff", |formatter| {
+        let tracked = tree
+            .entries()
+            .map(|entry| entry.0.into_internal_string())
+            .collect::<BTreeSet<String>>();
```

Comment:
It can be super expensive to collect all tree entries recursively. Maybe we can instead visit trees from the root by using `MergedTree::sub_tree(name)` or `Tree::value(name)`, and cache visited entries.

Reply:
Did this and it seems to work. It's slightly more complex which makes me think it may be conceptually simpler if actually done while snapshotting, so that's another reason to do it properly in the future.
(also I changed emptiness to do `tree.names().next()` since that seems like it would be cheaper)